File: src/main/java/cn/labzen/web/ap/processor/EvaluateFieldsProcessor.java
package cn.labzen.web.ap.processor;

import cn.labzen.web.ap.internal.Utils;
import cn.labzen.web.ap.internal.context.ControllerContext;
import cn.labzen.web.ap.internal.element.ElementAnnotation;
import cn.labzen.web.ap.internal.element.ElementClass;
import cn.labzen.web.ap.internal.element.ElementField;
import cn.labzen.web.ap.suggestion.AppendSuggestion;
import cn.labzen.web.ap.suggestion.RemoveSuggestion;
import cn.labzen.web.ap.suggestion.ReplaceSuggestion;
import cn.labzen.web.ap.suggestion.Suggestion;
import com.squareup.javapoet.TypeName;

import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import java.util.List;
import java.util.stream.Collectors;

public final class EvaluateFieldsProcessor implements InternalProcessor {

  @Override
  public void process(ControllerContext context) {
    // 读取所有继承的父接口定义的泛型参数类型
    List<? extends TypeMirror> directSupertypes = context.getApc().getTypes().directSupertypes(context.getSource().asType());

    directSupertypes.forEach(inter -> {
      List<TypeName> typeArguments = detectTypeArguments(inter);
      if (typeArguments == null) return;

      TypeName interfaceClassName = Utils.typeOf(inter);

      // 遍历每一个评价器
      List<Suggestion> suggestions = context.getGenericsEvaluators().stream()
        .filter(evaluator -> evaluator.support(interfaceClassName))
        .flatMap(evaluator -> evaluator.evaluate(typeArguments).stream())
        .collect(Collectors.toList());

      suggestions.forEach(suggestion -> {
        switch (suggestion) {
          case AppendSuggestion append -> parseAppendSuggestion(context.getRoot(), append);
          case RemoveSuggestion remove -> parseRemoveSuggestion(context.getRoot(), remove);
          case ReplaceSuggestion replace -> parseReplaceSuggestion(context.getRoot(), replace);
          default -> throw new IllegalStateException("Unexpected suggestion: " + suggestion);
        }
      });
    });
  }

  private void parseAppendSuggestion(ElementClass root, AppendSuggestion suggestion) {
    if (suggestion.element() instanceof ElementField field) {
      root.getFields().add(field);
    } else if (suggestion.element() instanceof ElementAnnotation annotation) {
      if (ElementClass.class.equals(suggestion.kind())) {
        root.getAnnotations().add(annotation);
      }
    }
  }

  private void parseRemoveSuggestion(ElementClass root, RemoveSuggestion suggestion) {
    if (ElementClass.class.equals(suggestion.kind())) {
      return;
    }

    List<ElementField> needlessFields = root.getFields().stream().filter(filed -> filed.keyword().equals(suggestion.keyword())).toList();
    needlessFields.forEach(field -> root.getFields().remove(field));
    List<ElementAnnotation> needlessAnnotations = root.getAnnotations().stream().filter(ann -> ann.keyword().equals(suggestion.keyword())).toList();
    needlessAnnotations.forEach(field -> root.getAnnotations().remove(field));
  }

  private void parseReplaceSuggestion(ElementClass root, ReplaceSuggestion suggestion) {
    if (suggestion.element() instanceof ElementAnnotation annotation) {
      List<ElementAnnotation> annotations = root.getAnnotations().stream().filter(ann -> ann.keyword().equals(suggestion.keyword())).toList();
      annotations.forEach(ann -> ann.getMembers().putAll(annotation.getMembers()));
    }
  }

  private List<TypeName> detectTypeArguments(TypeMirror typeMirror) {
    if (typeMirror instanceof DeclaredType declaredType) {
      List<? extends TypeMirror> typeArguments = declaredType.getTypeArguments();
      return typeArguments.stream().map(Utils::typeOf).toList();
    } else {
      return null;
    }
  }

  @Override
  public int priority() {
    return PRIORITY_EVALUATE_FIELDS;
  }
}

File: src/main/kotlin/cn/labzen/web/ap/processor/EvaluateFieldsProcessor.kt
package cn.labzen.web.ap.processor

import cn.labzen.web.ap.internal.Utils
import cn.labzen.web.ap.internal.context.ControllerContext
import cn.labzen.web.ap.internal.element.ElementAnnotation
import cn.labzen.web.ap.internal.element.ElementClass
import cn.labzen.web.ap.internal.element.ElementField
import cn.labzen.web.ap.processor.InternalProcessor.Companion.PRIORITY_EVALUATE_FIELDS
import cn.labzen.web.ap.suggestion.impl.AppendSuggestion
import cn.labzen.web.ap.suggestion.impl.RemoveSuggestion
import cn.labzen.web.ap.suggestion.impl.ReplaceSuggestion
import com.squareup.javapoet.TypeName
import javax.lang.model.type.DeclaredType
import javax.lang.model.type.TypeMirror

/**
 * 解析Controller父接口声明的泛型，转换为Controller实现类中需要添加的Field
 */
class EvaluateFieldsProcessor : InternalProcessor {

  override fun process(context: ControllerContext) {
    // 读取所有继承的父接口定义的泛型参数类型
    val directSupertypes = context.apc.typeUtils.directSupertypes(context.source.asType())

    directSupertypes.forEach { inter ->
      val typeArguments = detectTypeArguments(inter) ?: return@forEach

      val interfaceClassName: TypeName = Utils.typeOf(inter)

      // 遍历每一个评价器
      val suggestions = context.genericsEvaluators.flatMap { evaluator ->
        if (evaluator.support(interfaceClassName)) {
          evaluator.evaluate(typeArguments)
        } else listOf()
      }

      suggestions.forEach { suggestion ->
        when (suggestion) {
          is AppendSuggestion -> parseAppendSuggestion(context.root, suggestion)
          is RemoveSuggestion -> parseRemoveSuggestion(context.root, suggestion)
          is ReplaceSuggestion -> parseReplaceSuggestion(context.root, suggestion)
        }
      }
    }
  }

  private fun parseAppendSuggestion(root: ElementClass, suggestion: AppendSuggestion) {
    when (suggestion.element) {
      is ElementField -> {
        root.fields.add(suggestion.element)
      }

      is ElementAnnotation -> {
        if (suggestion.kind == ElementClass::class.java) {
          root.annotations.add(suggestion.element)
        }
      }
    }
  }

  private fun parseRemoveSuggestion(root: ElementClass, suggestion: RemoveSuggestion) {
    if (suggestion.kind != ElementClass::class.java) {
      return
    }
    root.fields.removeIf {
      it.keyword() == suggestion.keyword
    }
    root.annotations.removeIf {
      it.keyword() == suggestion.keyword
    }
  }

  private fun parseReplaceSuggestion(root: ElementClass, suggestion: ReplaceSuggestion) {
    val element = suggestion.element
    when (element) {
      is ElementAnnotation -> {
        root.annotations.find {
          it.keyword() == suggestion.keyword
        }?.apply {
          members.putAll(element.members)
        }
      }
    }
  }

  private fun detectTypeArguments(typeMirror: TypeMirror): List<TypeName>? =
    if (typeMirror is DeclaredType) {
      val typeArguments = typeMirror.typeArguments
      typeArguments.map { Utils.typeOf(it) }
    } else null

  override fun priority(): Int = PRIORITY_EVALUATE_FIELDS

}